text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,707 @@
+        .provide_ref::<[&'static Location<'static>]>(&self.locations)
+        // can be used to upcast self to other trait objects
+        .provide_ref::<dyn Serialize>(&self)
+        // or to pass owned values
+        .provide_value::<ExitCode>(self.exit_code);
+}
+```
+
+Example usage:
+
+```rust
+let e: &dyn Error = &concrete_error;
+
+if let Some(bt) = e.context_ref::<Backtrace>() {
+    println!("{}", bt);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+In Rust, errors typically gather two forms of context when they are created:
+context for the *current error message* and context for the *final* *error
+report*. The `Error` trait exists to provide an interface to context intended
+for error reports. This context includes the error message, the source error,
+and, more recently, backtraces.
+
+However, the current approach of promoting each form of context to a method on
+the `Error` trait doesn't leave room for forms of context that are not commonly
+used, or forms of context that are defined outside of the standard library.
+
+## Extracting non-std types from `dyn Errors`
+
+By adding a generic form of these member access functions we are no longer
+restricted to types defined in the standard library. This opens the door to
+many new forms of error reporting.
+
+### Example use cases this enables
+
+* using alternatives to `std::backtrace::Backtrace` such as
+  `backtrace::Backtrace` or [`SpanTrace`]
+* zig-like Error Return Traces by extracting `Location` types from errors
+  gathered via `#[track_caller]` or similar.
+* error source trees instead of chains by accessing the source of an error as a
+  slice of errors rather than as a single error, such as a set of errors caused
+  when parsing a file TODO reword
+* Help text such as suggestions or warnings attached to an error report
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Error handling in Rust consists of three main steps: creation/propagation,
+handling, and reporting. The `std::error::Error` trait exists to bridge the
+gap between creation and reporting. It does so by acting as an interface that
+all error types can implement that defines how to access context intended for
+error reports, such as the error message, source, or location it was created.
+This allows error reporting types to handle errors in a consistent manner
+when constructing reports for end users while still retaining control over
+the format of the full report.
+
+The `Error` trait accomplishes this by providing a set of methods for accessing
+members of `dyn Error` trait objects. It requires that types implement the
+`Display` trait, which acts as the interface to the main member, the error
+message itself.  It provides the `source` function for accessing `dyn Error`
+members, which typically represent the current error's cause. Via
+`#![feature(backtrace)]` it provides the `backtrace` function, for accessing a
```

Comment:
This seems outdated. In master branch I don't see any `backtrace` function on the `Error` trait with or without that feature. Maybe the `backtrace` feature was already merged but left out the `backtrace` method on `Error`. This RFC makes it unnecessary anyway so I'll look into updating this section here.

Reply:
There used to be an unstable `backtrace` function on `Error`, but it was removed in favor for the generic member access proposed in this RFC.